In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
cd drive/'My Drive'/'Data mining'/bbc_text_categorization

/content/drive/My Drive/Data mining/bbc_text_categorization


In [ ]:
import pandas as pd
import numpy as np
import os
from sklearn.model_selection import train_test_split
from sklearn import metrics as sklm
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB, GaussianNB, BernoulliNB, CategoricalNB
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from pprint import pprint
from time import time
from sklearn.preprocessing import FunctionTransformer

In [ ]:
DATA_PATH='data/bbc-text.csv'

## Data overview

In [ ]:
df=pd.read_csv(DATA_PATH)
df.sample(10)

category                                               text
883        politics  brown in appeal for labour unity gordon brown ...
1989          sport  hodgson shoulders england blame fly-half charl...
1587          sport  wenger signs new deal arsenal manager arsene w...
2179  entertainment  musicians to tackle us red tape musicians  gro...
1867          sport  federer breezes into semi-finals roger federer...
1534  entertainment  oscar nominees gear up for lunch leonardo dica...
356        politics  green fear for transport ballot the green part...
36             tech  gamers snap up new sony psp gamers have bought...
1629       business  wmc says xstrata bid is too low australian min...
468        politics  guantanamo four free in weeks all four britons...

In [ ]:
df.describe(include='all')

category                                               text
count      2225                                               2225
unique        5                                               2126
top       sport  queen recruit singer for new tour the remainin...
freq        511                                                  2

Note: There are some duplicate samples in data: 2225 sample but 2126 unique text values

In [ ]:
df.drop_duplicates(inplace=True)
df.nunique()

category       5
text        2126
dtype: int64

In [ ]:
df['category'].value_counts()

sport            504
business         503
politics         403
entertainment    369
tech             347
Name: category, dtype: int64

In [ ]:
print('Example sample:')
print(f"- Label: {df['category'][0]}")
print('- Content:', df['text'][0], sep='\n')

Example sample:
- Label: tech
- Content:
tv future in the hands of viewers with home theatre systems  plasma high-definition tvs  and digital video recorders moving into the living room  the way people watch tv will be radically different in five years  time.  that is according to an expert panel which gathered at the annual consumer electronics show in las vegas to discuss how these new technologies will impact one of our favourite pastimes. with the us leading the trend  programmes and other content will be delivered to viewers via home networks  through cable  satellite  telecoms companies  and broadband service providers to front rooms and portable devices.  one of the most talked-about technologies of ces has been digital and personal video recorders (dvr and pvr). these set-top boxes  like the us s tivo and the uk s sky+ system  allow people to record  store  play  pause and forward wind tv programmes when they want.  essentially  the technology allows for much more personalised 

In [ ]:
CATEGORIES=df['category'].unique().tolist()
print(CATEGORIES)

['tech', 'business', 'sport', 'entertainment', 'politics']


In [ ]:
X=df['text'].to_numpy()
Y=df['category'].to_numpy()

In [ ]:
np.unique(Y)

array(['business', 'entertainment', 'politics', 'sport', 'tech'],
      dtype=object)

## Train test split

In [ ]:
# Test set of 500 samples
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=500, random_state=42)
print(x_train.shape, y_train.shape, x_test.shape, y_test.shape)
# Test set 2, a subset of test set contains only 300 samples
x_test2, _, y_test2, __= train_test_split(x_test, y_test, test_size=300, random_state=42)
print(x_test2.shape, y_test2.shape)

(1626,) (1626,) (500,) (500,)
(200,) (200,)


## Metrics evaluation
Function to evaluate some metrics

In [ ]:
def evaluate_metrics(y_true, y_pred, labels=None):
    metrics = sklm.precision_recall_fscore_support(y_true, y_pred)
    cm = sklm.confusion_matrix(y_true, y_pred, labels=labels)
    num_classes=cm.shape[0]
    
    # For each class, calculate metrics
    metrics=pd.DataFrame(columns=['support', 'True Positive Rate(TPR)', 'False Positive Rate(FPR)',
        'Sensitivity(SE)','Specificity(SP)', 'Accuracy(ACC)', 'Precision(P)','Recall(R)','F1-Score'])
    n_sample=np.sum(cm)
    for i in range(num_classes):
        # True positive rate, False Positive rate, Sensitivity, Specificity
        support=np.sum(cm[i,:])
        tp=cm[i,i]
        fn=support - tp
        fp=np.sum(cm[:, i])-tp
        tn=n_sample-tp-fn-fp
        tpr=tp/(tp+fn)
        fpr=fp/(fp+tn)
        se=tpr
        sp=1-fpr
        acc=(tp+tn)/n_sample
        precision=tp/(tp+fp)
        recall=tpr
        f1_s=2*precision*recall/(precision+recall)
        metrics.loc[labels[i]]=[support, tpr, fpr, se, sp, acc, precision, recall, f1_s]
    return cm, metrics

## Binary classification
Testing with binary classification problem: sport vs non-sport labels

### Transform labels to binary classification
Tranform label array for binary classes: sport vs non-sport

In [ ]:
y_test_bin=y_test.copy()
y_test2_bin=y_test2.copy()

for arr in [y_train_bin, y_test_bin, y_test2_bin]:
    for i in range(arr.shape[0]):
        if arr[i] != 'sport':
            arr[i]='non-sport'
    # Verify
    print(arr.shape)
    print(np.unique(arr))
    print('----------')

(1626,)
['non-sport' 'sport']
----------
(500,)
['non-sport' 'sport']
----------
(200,)
['non-sport' 'sport']
----------


### Binary feature vector using CountVectorizer
Using CountVectorizer to construct binary features:   
https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html#sklearn.feature_extraction.text.CountVectorizer

In [ ]:
vectorizer=CountVectorizer(lowercase=True, analyzer='word', binary=True, stop_words='english', max_df=0.75, max_features=1000, ngram_range=(1,2))
x_train_bin=vectorizer.fit_transform(x_train).todense()
print(x_train_bin.shape)
print(x_train_bin)

(1626, 1000)
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [1 0 0 ... 0 0 0]
 [1 0 0 ... 0 0 0]]


In [ ]:
print('Stopwords:', vectorizer.get_stop_words())
print('Vocabulary:', vectorizer.get_feature_names())

Stopwords: frozenset({'themselves', 'over', 'thin', 'might', 'whereupon', 'for', 'us', 'becoming', 'something', 'whither', 'back', 'do', 'interest', 'once', 'they', 'my', 'cant', 'here', 'con', 'beyond', 'her', 'done', 'meanwhile', 'side', 'whereafter', 'or', 'someone', 'besides', 'well', 'bottom', 'sincere', 'thru', 'without', 'almost', 'per', 'seems', 'yourself', 'least', 'fifteen', 'anywhere', 'four', 'hereby', 'latter', 'serious', 'himself', 'top', 'myself', 'towards', 'three', 'that', 'detail', 'can', 'thereafter', 'via', 'am', 'nothing', 'much', 'now', 'rather', 'a', 'still', 'we', 'any', 'latterly', 'anything', 'front', 'wherever', 're', 'again', 'thereupon', 'how', 'nevertheless', 'this', 'between', 'de', 'each', 'him', 'always', 'across', 'often', 'ours', 'would', 'both', 'found', 'off', 'less', 'herein', 'everywhere', 'his', 'some', 'the', 'everything', 'becomes', 'indeed', 'are', 'never', 'see', 'since', 'though', 'had', 'what', 'as', 'whereas', 'which', 'at', 'afterwards', 

In [ ]:
x_train_bin[:1]

matrix([[0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
         0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0

In [ ]:
x_test_bin=vectorizer.transform(x_test).todense()
x_test2_bin=vectorizer.transform(x_test2).todense()
print(x_test_bin.shape, x_test2_bin.shape)

(500, 1000) (200, 1000)


#### Which Naive Bayes model performs better   
https://scikit-learn.org/stable/modules/naive_bayes.html

In [ ]:
def train_and_test(clf, x_train, y_train, x_test, y_test):
    clf.fit(x_train_bin, y_train_bin)
    print(clf)
    print(f'Train set accuracy: {clf.score(x_train_bin, y_train_bin)}')
    print(f'Test set accuracy: {clf.score(x_test_bin, y_test_bin)}')

In [ ]:
clfs=[MultinomialNB(), BernoulliNB(), CategoricalNB()]

BernoulliNB and CategoricalNB should produces the same result since features in data are all in binary form

In [ ]:
for clf in clfs:
    train_and_test(clf, x_train_bin, y_train_bin, x_test_bin, y_test_bin)
    print('---------------------')

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)
Train set accuracy: 0.992619926199262
Test set accuracy: 0.996
---------------------
BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)
Train set accuracy: 0.9895448954489545
Test set accuracy: 0.992
---------------------
CategoricalNB(alpha=1.0, class_prior=None, fit_prior=True)
Train set accuracy: 0.9895448954489545
Test set accuracy: 0.992
---------------------


In [ ]:
clf=MultinomialNB()
clf.fit(x_train_bin, y_train_bin)
y_test_pred=clf.predict(x_test_bin)
y_test2_pred=clf.predict(x_test2_bin)

##### Test set of 500 samples

In [ ]:
cm, metrics= evaluate_metrics(y_test_bin, y_test_pred, labels=['sport', 'non-sport'])
print('Confusion matrix')
print(cm)
metrics

Confusion matrix
[[117   2]
 [  0 381]]


support  True Positive Rate(TPR)  False Positive Rate(FPR)  \
sport        119.0                 0.983193                  0.000000   
non-sport    381.0                 1.000000                  0.016807   

           Sensitivity(SE)  Specificity(SP)  Accuracy(ACC)  Precision(P)  \
sport             0.983193         1.000000          0.996      1.000000   
non-sport         1.000000         0.983193          0.996      0.994778   

           Recall(R)  F1-Score  
sport       0.983193  0.991525  
non-sport   1.000000  0.997382

Verify result metrics, our evaluate_metrics() funtion should produce the same result as sklearn 's classification_report() function

In [ ]:
report = sklm.classification_report(y_test_bin, y_test_pred, labels= ['sport', 'non-sport'])
print(report)

              precision    recall  f1-score   support

       sport       1.00      0.98      0.99       119
   non-sport       0.99      1.00      1.00       381

    accuracy                           1.00       500
   macro avg       1.00      0.99      0.99       500
weighted avg       1.00      1.00      1.00       500



##### Test set of 300 samples

In [ ]:
cm2, metrics2= evaluate_metrics(y_test2_bin, y_test2_pred, labels=['sport', 'non-sport'])
print('Confusion matrix')
print(cm2)
metrics2

Confusion matrix
[[ 52   2]
 [  0 146]]


support  True Positive Rate(TPR)  False Positive Rate(FPR)  \
sport         54.0                 0.962963                  0.000000   
non-sport    146.0                 1.000000                  0.037037   

           Sensitivity(SE)  Specificity(SP)  Accuracy(ACC)  Precision(P)  \
sport             0.962963         1.000000           0.99      1.000000   
non-sport         1.000000         0.962963           0.99      0.986486   

           Recall(R)  F1-Score  
sport       0.962963  0.981132  
non-sport   1.000000  0.993197

### TF-IDF features vector using TfidfVectorizer

In [ ]:
vectorizer=TfidfVectorizer(lowercase=True, analyzer='word', stop_words='english', max_df=0.75,
                           max_features=1000, ngram_range=(1,2), use_idf=True, smooth_idf=True)
x_train_tfidf=vectorizer.fit_transform(x_train).todense()
print(x_train_tfidf.shape)
print(x_train_tfidf)

(1626, 1000)
[[0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 ...
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.13231895 0.         0.         ... 0.         0.         0.        ]
 [0.07246767 0.         0.         ... 0.         0.         0.        ]]


In [ ]:
print('Stopwords:', vectorizer.get_stop_words())
print('Vocabulary:', vectorizer.get_feature_names())

Stopwords: frozenset({'themselves', 'over', 'thin', 'might', 'whereupon', 'for', 'us', 'becoming', 'something', 'whither', 'back', 'do', 'interest', 'once', 'they', 'my', 'cant', 'here', 'con', 'beyond', 'her', 'done', 'meanwhile', 'side', 'whereafter', 'or', 'someone', 'besides', 'well', 'bottom', 'sincere', 'thru', 'without', 'almost', 'per', 'seems', 'yourself', 'least', 'fifteen', 'anywhere', 'four', 'hereby', 'latter', 'serious', 'himself', 'top', 'myself', 'towards', 'three', 'that', 'detail', 'can', 'thereafter', 'via', 'am', 'nothing', 'much', 'now', 'rather', 'a', 'still', 'we', 'any', 'latterly', 'anything', 'front', 'wherever', 're', 'again', 'thereupon', 'how', 'nevertheless', 'this', 'between', 'de', 'each', 'him', 'always', 'across', 'often', 'ours', 'would', 'both', 'found', 'off', 'less', 'herein', 'everywhere', 'his', 'some', 'the', 'everything', 'becomes', 'indeed', 'are', 'never', 'see', 'since', 'though', 'had', 'what', 'as', 'whereas', 'which', 'at', 'afterwards', 

In [ ]:
x_train_tfidf[:1]

matrix([[0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.05408359,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.03378191,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0

In [ ]:
x_test_tfidf=vectorizer.transform(x_test).todense()
x_test2_tfidf=vectorizer.transform(x_test2).todense()
print(x_test_tfidf.shape, x_test2_tfidf.shape)

(500, 1000) (200, 1000)


#### Which Naive Bayes model performs better   
https://scikit-learn.org/stable/modules/naive_bayes.html  
GaussianNB and MultinomialNB may work well with TF-IDF features vector

In [ ]:
clfs=[GaussianNB(), MultinomialNB()]
for clf in clfs:
    train_and_test(clf, x_train_tfidf, y_train_bin, x_test_tfidf, y_test_bin)
    print('---------------------')

GaussianNB(priors=None, var_smoothing=1e-09)
Train set accuracy: 0.986469864698647
Test set accuracy: 0.98
---------------------
MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)
Train set accuracy: 0.992619926199262
Test set accuracy: 0.996
---------------------


In [ ]:
clf=MultinomialNB()
clf.fit(x_train_tfidf, y_train_bin)
y_test_pred=clf.predict(x_test_tfidf)
y_test2_pred=clf.predict(x_test2_tfidf)

##### Test set of 500 samples

In [ ]:
cm, metrics= evaluate_metrics(y_test_bin, y_test_pred, labels=['sport', 'non-sport'])
print('Confusion matrix')
print(cm)
metrics

Confusion matrix
[[118   1]
 [  0 381]]


support  True Positive Rate(TPR)  False Positive Rate(FPR)  \
sport        119.0                 0.991597                  0.000000   
non-sport    381.0                 1.000000                  0.008403   

           Sensitivity(SE)  Specificity(SP)  Accuracy(ACC)  Precision(P)  \
sport             0.991597         1.000000          0.998      1.000000   
non-sport         1.000000         0.991597          0.998      0.997382   

           Recall(R)  F1-Score  
sport       0.991597  0.995781  
non-sport   1.000000  0.998689

##### Test set of 300 samples

In [ ]:
cm2, metrics2= evaluate_metrics(y_test2_bin, y_test2_pred, labels=['sport', 'non-sport'])
print('Confusion matrix')
print(cm2)
metrics2

Confusion matrix
[[ 53   1]
 [  0 146]]


support  True Positive Rate(TPR)  False Positive Rate(FPR)  \
sport         54.0                 0.981481                  0.000000   
non-sport    146.0                 1.000000                  0.018519   

           Sensitivity(SE)  Specificity(SP)  Accuracy(ACC)  Precision(P)  \
sport             0.981481         1.000000          0.995      1.000000   
non-sport         1.000000         0.981481          0.995      0.993197   

           Recall(R)  F1-Score  
sport       0.981481  0.990654  
non-sport   1.000000  0.996587

## Multi-classes classification (5 classes)

### Binary features vector

#### Build a pipeline to find out best hyper parameters

In [ ]:
def grid_search(pipeline, parameters, x_train, y_train):
    # find the best parameters for both the feature extraction and the classifier
    grid_searcher = GridSearchCV(pipeline, parameters, scoring='accuracy',
                               refit='accuracy', n_jobs=1,  verbose=10, cv=3)
    print("Performing grid search...")
    print("pipeline:", [name for name, _ in pipeline.steps])
    print("parameters:")
    pprint(parameters)
    t0 = time()
    grid_searcher.fit(x_train, y_train)
    print("done in %0.3fs" % (time() - t0))
    print("Best score: %0.3f" % grid_searcher.best_score_)
    print("Best parameters set:")
    best_estimator=grid_searcher.best_estimator_
    best_parameters = best_estimator.get_params()
    print(best_parameters)
    return grid_searcher, best_estimator

In [ ]:
parameters = {
    'vectorizer__stop_words': (None, 'english'),
    'vectorizer__max_features': (100, 300, 1000, 3000, 10000),
    'vectorizer__ngram_range': ((1, 1), (1, 2), (1,3)),
}

In [ ]:
bernoulli_pipeline = Pipeline([
    ('vectorizer', CountVectorizer(lowercase=True, analyzer='word', binary=True, max_df=0.75)),
    ('todense', FunctionTransformer(lambda x: x.todense(), accept_sparse=True)),
    ('clf', BernoulliNB())
])
bernoulli_searcher, best_bernoulli_clf=grid_search(bernoulli_pipeline, parameters, x_train, y_train)

Performing grid search...
pipeline: ['vectorizer', 'todense', 'clf']
parameters:
{'vectorizer__max_features': (100, 300, 1000, 3000, 10000),
 'vectorizer__ngram_range': ((1, 1), (1, 2), (1, 3)),
 'vectorizer__stop_words': (None, 'english')}
Fitting 3 folds for each of 30 candidates, totalling 90 fits
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, score=0.661, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.5s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, score=0.703, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    1.0s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, score=0.673, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english 


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    1.5s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, score=0.799, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english 


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    2.0s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, score=0.762, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english 


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    2.5s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, score=0.815, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:    3.0s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None, score=0.648, total=   2.0s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:    5.0s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None, score=0.679, total=   1.9s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:    6.9s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None, score=0.659, total=   1.9s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english 


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:    8.9s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english, score=0.797, total=   1.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english, score=0.760, total=   1.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english, score=0.803, total=   1.6s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 3), vectorizer__stop_words=None 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 3), vectorizer__stop_words=None, score=0.648, total=   4.3s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 3), vectorizer__stop_words=None 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 3), vectorizer__stop_words=None, score=0.

[Parallel(n_jobs=1)]: Done  90 out of  90 | elapsed:  3.1min finished


done in 186.111s
Best score: 0.958
Best parameters set:
{'memory': None, 'steps': [('vectorizer', CountVectorizer(analyzer='word', binary=True, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=0.75, max_features=10000, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words='english',
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)), ('todense', FunctionTransformer(accept_sparse=True, check_inverse=True,
                    func=<function <lambda> at 0x7f6f3ebb1268>,
                    inv_kw_args=None, inverse_func=None, kw_args=None,
                    validate=False)), ('clf', BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True))], 'verbose': False, 'vectorizer': CountVectorizer(analyzer='word', binary=True, decode_error='strict',
                dtype=<class 'numpy.int64'>, encodi

In [ ]:
multinomial_pipeline = Pipeline([
    ('vectorizer', CountVectorizer(lowercase=True, analyzer='word', binary=True, max_df=0.75)),
    ('todense', FunctionTransformer(lambda x: x.todense(), accept_sparse=True)),
    ('clf', MultinomialNB())
])
multinomial_searcher, best_multinomial_clf=grid_search(multinomial_pipeline, parameters, x_train, y_train)

Performing grid search...
pipeline: ['vectorizer', 'todense', 'clf']
parameters:
{'vectorizer__max_features': (100, 300, 1000, 3000, 10000),
 'vectorizer__ngram_range': ((1, 1), (1, 2), (1, 3)),
 'vectorizer__stop_words': (None, 'english')}
Fitting 3 folds for each of 30 candidates, totalling 90 fits
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, score=0.729, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.5s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, score=0.727, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    1.0s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, score=0.731, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english 


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    1.5s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, score=0.812, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english 


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    2.0s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, score=0.810, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english 


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    2.5s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, score=0.810, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:    2.9s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None, score=0.725, total=   2.0s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:    4.9s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None, score=0.720, total=   1.9s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None 


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:    6.8s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None, score=0.701, total=   1.9s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english 


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:    8.8s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english, score=0.810, total=   1.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english, score=0.812, total=   1.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=english, score=0.797, total=   1.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 3), vectorizer__stop_words=None 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 3), vectorizer__stop_words=None, score=0.725, total=   4.2s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 3), vectorizer__stop_words=None 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 3), vectorizer__stop_words=None, score=0.

[Parallel(n_jobs=1)]: Done  90 out of  90 | elapsed:  3.0min finished


done in 185.893s
Best score: 0.964
Best parameters set:
{'memory': None, 'steps': [('vectorizer', CountVectorizer(analyzer='word', binary=True, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=0.75, max_features=10000, min_df=1,
                ngram_range=(1, 3), preprocessor=None, stop_words='english',
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)), ('todense', FunctionTransformer(accept_sparse=True, check_inverse=True,
                    func=<function <lambda> at 0x7f6f3de96f28>,
                    inv_kw_args=None, inverse_func=None, kw_args=None,
                    validate=False)), ('clf', MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))], 'verbose': False, 'vectorizer': CountVectorizer(analyzer='word', binary=True, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', 

In [ ]:
pd.set_option('display.max_columns', 100)
pd.DataFrame(multinomial_searcher.cv_results_).sort_values('rank_test_score')

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
29       2.724626      0.047175         0.321136        0.002090   
27       1.553121      0.007609         0.247730        0.005393   
25       0.583247      0.006003         0.165784        0.000643   
24       0.590531      0.002407         0.182325        0.004664   
21       1.398708      0.011437         0.218820        0.002411   
19       0.436595      0.009940         0.141678        0.002994   
18       0.442474      0.003588         0.153002        0.001319   
26       1.888304      0.006479         0.340842        0.005894   
23       2.578952      0.014529         0.304348        0.008537   
15       1.390334      0.013922         0.201407        0.001680   
17       2.561956      0.022277         0.279140        0.002650   
28       3.961259      0.049500         0.493210        0.003727   
22       3.836608      0.030969         0.452834        0.016013   
13       0.394583      0.002407         0.122006        0.000782   
20       1.741129      0.011977         0.296297        0.002243   
12       0.407549      0.015627         0.137296        0.003751   
14       1.705050      0.018799         0.281391        0.012287   
16       3.833656      0.049469         0.420495        0.003603   
9        1.345438      0.006918         0.190742        0.001500   
11       2.543753      0.025408         0.265659        0.002382   
7        0.366893      0.002503         0.112873        0.000669   
6        0.389053      0.011306         0.124183        0.000708   
10       3.849944      0.044803         0.408031        0.003355   
8        1.663649      0.018367         0.256053        0.001550   
1        0.365110      0.001481         0.110610        0.001779   
5        2.521178      0.031578         0.259550        0.006071   
3        1.343870      0.012469         0.188307        0.004511   
0        0.372681      0.017201         0.123043        0.003670   
4        3.775060      0.060636         0.394406        0.006579   
2        1.684870      0.024895         0.249874        0.001119   

   param_vectorizer__max_features param_vectorizer__ngram_range  \
29                          10000                        (1, 3)   
27                          10000                        (1, 2)   
25                          10000                        (1, 1)   
24                          10000                        (1, 1)   
21                           3000                        (1, 2)   
19                           3000                        (1, 1)   
18                           3000                        (1, 1)   
26                          10000                        (1, 2)   
23                           3000                        (1, 3)   
15                           1000                        (1, 2)   
17                           1000                        (1, 3)   
28                          10000                        (1, 3)   
22                           3000                        (1, 3)   
13                           1000                        (1, 1)   
20                           3000                        (1, 2)   
12                           1000                        (1, 1)   
14                           1000                        (1, 2)   
16                           1000                        (1, 3)   
9                             300                        (1, 2)   
11                            300                        (1, 3)   
7                             300                        (1, 1)   
6                             300                        (1, 1)   
10                            300                        (1, 3)   
8                             300                        (1, 2)   
1                             100                        (1, 1)   
5                             100                        (1, 3)   
3                             100                        (1, 2)   
0                  

#### Evaluate metrics

##### Test set of 500 samples

In [ ]:
print('Accuracy:', best_multinomial_clf.score(x_test, y_test))
y_test_pred=best_multinomial_clf.predict(x_test)
cm, metric_df=evaluate_metrics(y_test, y_test_pred, labels=CATEGORIES)
print('Confusion matrix:')
print(cm)
metric_df

Accuracy: 0.97
Confusion matrix:
[[ 81   5   0   2   0]
 [  3 113   0   0   1]
 [  0   0 119   0   0]
 [  2   0   0  77   0]
 [  0   2   0   0  95]]


support  True Positive Rate(TPR)  False Positive Rate(FPR)  \
tech              88.0                 0.920455                  0.012136   
business         117.0                 0.965812                  0.018277   
sport            119.0                 1.000000                  0.000000   
entertainment     79.0                 0.974684                  0.004751   
politics          97.0                 0.979381                  0.002481   

               Sensitivity(SE)  Specificity(SP)  Accuracy(ACC)  Precision(P)  \
tech                  0.920455         0.987864          0.976      0.941860   
business              0.965812         0.981723          0.978      0.941667   
sport                 1.000000         1.000000          1.000      1.000000   
entertainment         0.974684         0.995249          0.992      0.974684   
politics              0.979381         0.997519          0.994      0.989583   

               Recall(R)  F1-Score  
tech            0.920455  0.931034  
business        0.965812  0.953586  
sport           1.000000  1.000000  
entertainment   0.974684  0.974684  
politics        0.979381  0.984456

##### Test set of 300 sample

In [ ]:
print('Accuracy:', best_multinomial_clf.score(x_test2, y_test2))
y_test2_pred=best_multinomial_clf.predict(x_test2)
cm, metric_df=evaluate_metrics(y_test2, y_test2_pred, labels=CATEGORIES)
print('Confusion matrix:')
print(cm)
metric_df

Accuracy: 0.97
Confusion matrix:
[[34  3  0  0  0]
 [ 1 41  0  0  1]
 [ 0  0 54  0  0]
 [ 0  0  0 28  0]
 [ 0  1  0  0 37]]


support  True Positive Rate(TPR)  False Positive Rate(FPR)  \
tech              37.0                 0.918919                  0.006135   
business          43.0                 0.953488                  0.025478   
sport             54.0                 1.000000                  0.000000   
entertainment     28.0                 1.000000                  0.000000   
politics          38.0                 0.973684                  0.006173   

               Sensitivity(SE)  Specificity(SP)  Accuracy(ACC)  Precision(P)  \
tech                  0.918919         0.993865           0.98      0.971429   
business              0.953488         0.974522           0.97      0.911111   
sport                 1.000000         1.000000           1.00      1.000000   
entertainment         1.000000         1.000000           1.00      1.000000   
politics              0.973684         0.993827           0.99      0.973684   

               Recall(R)  F1-Score  
tech            0.918919  0.944444  
business        0.953488  0.931818  
sport           1.000000  1.000000  
entertainment   1.000000  1.000000  
politics        0.973684  0.973684

### TF-IDF features vector

#### Build a pipeline to find out best hyper parameters

In [ ]:
parameters = {
    'vectorizer__stop_words': (None, 'english'),
    'vectorizer__max_features': (100, 300, 1000, 3000, 10000),
    'vectorizer__ngram_range': ((1, 1), (1, 2), (1,3)), 
    'vectorizer__sublinear_tf':(True, False)
}

In [ ]:
gaussian_pipeline = Pipeline([
    ('vectorizer', TfidfVectorizer(lowercase=True, analyzer='word', use_idf=True, smooth_idf=True, max_df=0.75)),
    ('todense', FunctionTransformer(lambda x: x.todense(), accept_sparse=True)),
    ('clf', GaussianNB())
])
gaussian_searcher, best_bernoulli_clf = grid_search(gaussian_pipeline, parameters, x_train, y_train)

Performing grid search...
pipeline: ['vectorizer', 'todense', 'clf']
parameters:
{'vectorizer__max_features': (100, 300, 1000, 3000, 10000),
 'vectorizer__ngram_range': ((1, 1), (1, 2), (1, 3)),
 'vectorizer__stop_words': (None, 'english'),
 'vectorizer__sublinear_tf': (True, False)}
Fitting 3 folds for each of 60 candidates, totalling 180 fits
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True, score=0.795, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.5s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True, score=0.782, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    1.0s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True, score=0.753, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False 


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    1.4s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False, score=0.808, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False 


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    1.9s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False, score=0.792, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False 


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    2.4s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False, score=0.742, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:    2.8s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True, score=0.793, total=   0.4s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:    3.3s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True, score=0.814, total=   0.4s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:    3.7s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True, score=0.821, total=   0.4s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False 


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:    4.1s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False, score=0.827, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False, score=0.815, total=   0.4s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False, score=0.817, total=   0.4s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None, vectorizer__sublinear_tf=True 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None, vectorizer__sublinear_tf=True, score=

[Parallel(n_jobs=1)]: Done 180 out of 180 | elapsed:  6.2min finished


done in 372.440s
Best score: 0.949
Best parameters set:
{'memory': None, 'steps': [('vectorizer', TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.float64'>, encoding='utf-8',
                input='content', lowercase=True, max_df=0.75,
                max_features=10000, min_df=1, ngram_range=(1, 2), norm='l2',
                preprocessor=None, smooth_idf=True, stop_words='english',
                strip_accents=None, sublinear_tf=True,
                token_pattern='(?u)\\b\\w\\w+\\b', tokenizer=None, use_idf=True,
                vocabulary=None)), ('todense', FunctionTransformer(accept_sparse=True, check_inverse=True,
                    func=<function <lambda> at 0x7f9012fe5c80>,
                    inv_kw_args=None, inverse_func=None, kw_args=None,
                    validate=False)), ('clf', GaussianNB(priors=None, var_smoothing=1e-09))], 'verbose': False, 'vectorizer': TfidfVectorizer(analyzer='word', binary=False, dec

In [ ]:
multinomial_pipeline = Pipeline([
    ('vectorizer', TfidfVectorizer(lowercase=True, analyzer='word', use_idf=True, smooth_idf=True, max_df=0.75)),
    ('todense', FunctionTransformer(lambda x: x.todense(), accept_sparse=True)),
    ('clf', MultinomialNB())
])
multinomial_searcher, best_multinomial_clf=grid_search(multinomial_pipeline, parameters, x_train, y_train)

Performing grid search...
pipeline: ['vectorizer', 'todense', 'clf']
parameters:
{'vectorizer__max_features': (100, 300, 1000, 3000, 10000),
 'vectorizer__ngram_range': ((1, 1), (1, 2), (1, 3)),
 'vectorizer__stop_words': (None, 'english'),
 'vectorizer__sublinear_tf': (True, False)}
Fitting 3 folds for each of 60 candidates, totalling 180 fits
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True, score=0.838, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.5s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True, score=0.790, total=   0.6s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    1.1s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=True, score=0.806, total=   0.6s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False 


[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    1.7s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False, score=0.828, total=   0.6s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False 


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    2.3s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False, score=0.786, total=   0.6s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False 


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    2.8s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=None, vectorizer__sublinear_tf=False, score=0.786, total=   0.6s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:    3.4s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True, score=0.904, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:    4.0s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True, score=0.902, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True 


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:    4.5s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=True, score=0.876, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False 


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:    5.1s remaining:    0.0s


[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False, score=0.902, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False, score=0.891, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 1), vectorizer__stop_words=english, vectorizer__sublinear_tf=False, score=0.878, total=   0.5s
[CV] vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None, vectorizer__sublinear_tf=True 
[CV]  vectorizer__max_features=100, vectorizer__ngram_range=(1, 2), vectorizer__stop_words=None, vectorizer__sublinear_tf=True, score=

[Parallel(n_jobs=1)]: Done 180 out of 180 | elapsed:  6.3min finished


done in 378.601s
Best score: 0.970
Best parameters set:
{'memory': None, 'steps': [('vectorizer', TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.float64'>, encoding='utf-8',
                input='content', lowercase=True, max_df=0.75, max_features=3000,
                min_df=1, ngram_range=(1, 2), norm='l2', preprocessor=None,
                smooth_idf=True, stop_words='english', strip_accents=None,
                sublinear_tf=False, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, use_idf=True, vocabulary=None)), ('todense', FunctionTransformer(accept_sparse=True, check_inverse=True,
                    func=<function <lambda> at 0x7f901257ebf8>,
                    inv_kw_args=None, inverse_func=None, kw_args=None,
                    validate=False)), ('clf', MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))], 'verbose': False, 'vectorizer': TfidfVectorizer(analyzer='word', binary=False, decod

In [ ]:
pd.set_option('display.max_columns', 100)
pd.DataFrame(multinomial_searcher.cv_results_).sort_values('rank_test_score')

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
43       1.389153      0.006463         0.261968        0.000568   
47       2.802945      0.194739         0.364105        0.022864   
59       2.633256      0.016991         0.362647        0.002070   
55       1.421249      0.012934         0.288123        0.005057   
38       0.404653      0.003656         0.188619        0.002124   
58       2.642627      0.012092         0.366806        0.003350   
54       1.423741      0.008016         0.288487        0.004106   
39       0.396864      0.003677         0.184775        0.000552   
46       2.665120      0.096048         0.367154        0.034214   
42       1.394368      0.015838         0.266610        0.006961   
51       0.451401      0.004436         0.216632        0.002951   
36       0.409638      0.006735         0.201589        0.001999   
37       0.423702      0.004974         0.199909        0.002116   
50       0.455341      0.007344         0.214547        0.003505   
45       3.886429      0.024576         0.495372        0.001094   
53       1.775210      0.013776         0.387220        0.002044   
41       1.715944      0.010888         0.355143        0.005971   
57       3.930703      0.038180         0.528182        0.006247   
34       2.569945      0.031412         0.328111        0.003002   
27       0.744136      0.029236         0.374873        0.072393   
52       1.789451      0.033101         0.393129        0.007430   
26       0.378338      0.007692         0.176083        0.002579   
30       1.375665      0.024349         0.251673        0.003215   
40       1.732634      0.018733         0.348475        0.003435   
35       2.609358      0.038124         0.325746        0.002101   
31       1.378778      0.014942         0.247922        0.000745   
49       0.470056      0.001811         0.236215        0.009776   
44       3.895674      0.045284         0.495344        0.005965   
56       3.929372      0.042519         0.527656        0.006187   
32       3.891564      0.108704         0.464814        0.004605   
28       2.810902      0.194381         0.508158        0.139442   
25       0.394643      0.002701         0.185998        0.000875   
24       0.385265      0.009175         0.190182        0.003545   
48       0.468256      0.005711         0.230010        0.003794   
29       1.723727      0.026192         0.323418        0.002224   
33       3.824822      0.050085         0.469416        0.015068   
19       1.359802      0.003608         0.240204        0.003136   
22       2.549337      0.027125         0.315024        0.002318   
14       0.370482      0.002247         0.165295        0.001541   
23       2.569321      0.018158         0.317633        0.003716   
18       1.371156      0.017928         0.239155        0.002734   
12       0.379266      0.012915         0.178658        0.001376   
15       0.372990      0.002829         0.168427        0.002592   
13       0.390764      0.000758         0.176254        0.000435   
16       1.691460      0.009419         0.313983        0.001447   
20       3.883744      0.045163         0.455914        0.002876   
21       3.834634      0.028856         0.453328        0.005408   
17       1.691516      0.019890         0.313046        0.002892   
10       2.563821      0.049368         0.315645        0.000316   
2        0.378076      0.008279         0.164126        0.001304   
6        1.356887      0.022044         0.234905        0.002309   
3        0.372686      0.005616         0.165454        0.002383   
11       2.565413      0.026201         0.310956        0.001935   
7        1.348032      0.012243         0.234420        0.000824   
0        0.380851      0.023553         0.175599        0.001700   
1        0.401247      0.007103         0.172397        0.000876   
4        1.675355      0.017399         0.305559        0.001096   
8        3.829330      0.038757         0.444275        0.00

#### Evaluate metrics

##### Test set of 500 samples

In [ ]:
print('Accuracy:', best_multinomial_clf.score(x_test, y_test))
y_test_pred=best_multinomial_clf.predict(x_test)
cm, metric_df=evaluate_metrics(y_test, y_test_pred, labels=CATEGORIES)
print('Confusion matrix:')
print(cm)
metric_df

Accuracy: 0.978
Confusion matrix:
[[ 81   2   1   2   2]
 [  1 116   0   0   0]
 [  0   0 119   0   0]
 [  0   0   0  79   0]
 [  0   3   0   0  94]]


support  True Positive Rate(TPR)  False Positive Rate(FPR)  \
tech              88.0                 0.920455                  0.002427   
business         117.0                 0.991453                  0.013055   
sport            119.0                 1.000000                  0.002625   
entertainment     79.0                 1.000000                  0.004751   
politics          97.0                 0.969072                  0.004963   

               Sensitivity(SE)  Specificity(SP)  Accuracy(ACC)  Precision(P)  \
tech                  0.920455         0.997573          0.984      0.987805   
business              0.991453         0.986945          0.988      0.958678   
sport                 1.000000         0.997375          0.998      0.991667   
entertainment         1.000000         0.995249          0.996      0.975309   
politics              0.969072         0.995037          0.990      0.979167   

               Recall(R)  F1-Score  
tech            0.920455  0.952941  
business        0.991453  0.974790  
sport           1.000000  0.995816  
entertainment   1.000000  0.987500  
politics        0.969072  0.974093

In [ ]:
report = sklm.classification_report(y_test, y_test_pred, labels= ['tech', 'business','sport','entertainment', 'politics'])
print(report)

               precision    recall  f1-score   support

         tech       0.99      0.92      0.95        88
     business       0.96      0.99      0.97       117
        sport       0.99      1.00      1.00       119
entertainment       0.98      1.00      0.99        79
     politics       0.98      0.97      0.97        97

     accuracy                           0.98       500
    macro avg       0.98      0.98      0.98       500
 weighted avg       0.98      0.98      0.98       500



##### Test set of 300 sample

In [ ]:
print('Accuracy:', best_multinomial_clf.score(x_test2, y_test2))
y_test2_pred=best_multinomial_clf.predict(x_test2)
cm, metric_df=evaluate_metrics(y_test2, y_test2_pred, labels=CATEGORIES)
print('Confusion matrix:')
print(cm)
metric_df

Accuracy: 0.98
Confusion matrix:
[[35  1  1  0  0]
 [ 1 42  0  0  0]
 [ 0  0 54  0  0]
 [ 0  0  0 28  0]
 [ 0  1  0  0 37]]


support  True Positive Rate(TPR)  False Positive Rate(FPR)  \
tech              37.0                 0.945946                  0.006135   
business          43.0                 0.976744                  0.012739   
sport             54.0                 1.000000                  0.006849   
entertainment     28.0                 1.000000                  0.000000   
politics          38.0                 0.973684                  0.000000   

               Sensitivity(SE)  Specificity(SP)  Accuracy(ACC)  Precision(P)  \
tech                  0.945946         0.993865          0.985      0.972222   
business              0.976744         0.987261          0.985      0.954545   
sport                 1.000000         0.993151          0.995      0.981818   
entertainment         1.000000         1.000000          1.000      1.000000   
politics              0.973684         1.000000          0.995      1.000000   

               Recall(R)  F1-Score  
tech            0.945946  0.958904  
business        0.976744  0.965517  
sport           1.000000  0.990826  
entertainment   1.000000  1.000000  
politics        0.973684  0.986667